File: multipicker/src/main/java/com/edwardstock/multipicker/picker/ui/PickerActivity.java
package com.edwardstock.multipicker.picker.ui;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.media.MediaScannerConnection;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.os.Looper;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import com.edwardstock.multipicker.PickerConfig;
import com.edwardstock.multipicker.R;
import com.edwardstock.multipicker.R2;
import com.edwardstock.multipicker.data.Dir;
import com.edwardstock.multipicker.data.MediaFile;
import com.edwardstock.multipicker.internal.CameraHandler;
import com.edwardstock.multipicker.internal.MediaFileLoader;
import com.edwardstock.multipicker.internal.PickerSavePath;
import com.edwardstock.multipicker.internal.PickerUtils;
import com.edwardstock.multipicker.internal.helpers.DisplayHelper;
import com.edwardstock.multipicker.internal.views.ErrorView;
import com.edwardstock.multipicker.internal.widgets.GridSpacingItemDecoration;
import com.edwardstock.multipicker.picker.PickerConst;
import com.edwardstock.multipicker.picker.views.PickerPresenter;
import com.edwardstock.multipicker.picker.views.PickerView;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.util.Pair;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import butterknife.BindView;
import permissions.dispatcher.NeedsPermission;
import permissions.dispatcher.RuntimePermissions;
import timber.log.Timber;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * android-multipicker. 2018
 * @author Eduard Maximovich [edward.vstock@gmail.com]
 */
@RuntimePermissions
public abstract class PickerActivity extends AppCompatActivity implements PickerView, ErrorView {
    public final static int RESULT_ADD_FILE_TO_SELECTION = Activity.RESULT_FIRST_USER + 1;
    private final static String STATE_SCANNED_FILES_FIRST_TIME = "STATE_SCANNED_FILES_FIRST_TIME";
    public @BindView(R2.id.toolbar) Toolbar toolbar;
    public @Nullable @BindView(R2.id.container_swipe_refresh) SwipeRefreshLayout swipeRefreshLayout;
    private PickerConfig mConfig;
    private GridSpacingItemDecoration mGridSpacingItemDecoration;
    protected static boolean scannedDirs = false;

    public PickerConfig getConfig() {
        if (mConfig == null) {
            final File in = new File(getCacheDir(), DirsActivity.CONFIG_FILE_NAME);

            try (RandomAccessFile raf = new RandomAccessFile(in, "r")) {
                byte[] cfgData = new byte[(int) in.length()];
                raf.readFully(cfgData);
                String cfgJson = new String(cfgData);
                Timber.d("Read json config: %s", cfgJson);
                mConfig = PickerConfig.fromJson(cfgJson);
            } catch (FileNotFoundException e) {
                throw new RuntimeException(e);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
        return mConfig;
    }

    @Override
    public void capturePhotoWithPermissions(CameraHandler cameraHandler, int requestCode) {
        PickerActivityPermissionsDispatcher.capturePhotoWithPermissionCheck(this, cameraHandler, requestCode);
    }

    @Override
    public void captureVideoWithPermissions(CameraHandler cameraHandler, int requestCode) {
        PickerActivityPermissionsDispatcher.captureVideoWithPermissionCheck(this, cameraHandler, requestCode);
    }

    @Override
    public void finishCapturePhoto(CameraHandler cameraHandler, Intent intent) {
//        mFragment.updateFiles();
    }

    @Override
    public void finishCaptureVideo(CameraHandler cameraHandler, Intent intent) {
//        mFragment.updateFiles();
    }

    @Override
    public void scanMedia(MediaFile file, MediaScannerConnection.OnScanCompletedListener listener) {
        String[] paths = {file.getPath()};
        String[] mimeTypes = {(file.isVideo() ? "video/mp4" : "image/jpeg")};
        MediaScannerConnection.scanFile(getApplicationContext(), paths, mimeTypes, listener);
    }

    @Override
    public void startUpdateFiles() {
        getPresenter().updateFiles(new MediaFileLoader(this));
    }

    @Override
    public void rescanFiles() {
        rescanFiles(null);
    }

    static <T> List<List<T>> chopped(List<T> list, final int L) {
        List<List<T>> parts = new ArrayList<>();
        final int N = list.size();
        for (int i = 0; i < N; i += L) {
            parts.add(new ArrayList<>(
                    list.subList(i, Math.min(N, i + L)))
            );
        }
        return parts;
    }

    @Override
    public void onError(CharSequence error) {
        Toast.makeText(this, error, Toast.LENGTH_LONG).show();
    }

    @Override
    public void onError(Throwable t) {
        Toast.makeText(this, t.getMessage(), Toast.LENGTH_LONG).show();
    }

    public void setupToolbar(@NonNull final Toolbar toolbar) {
        checkNotNull(toolbar, "Toolbar can't be null!");
        setSupportActionBar(toolbar);

        assert (getSupportActionBar() != null);
        getSupportActionBar().setHomeButtonEnabled(true);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        }

        if (item.getItemId() == R.id.menu_camera) {
            getPresenter().handleCapturePhoto();
            return true;
        } else if (item.getItemId() == R.id.menu_video) {
            getPresenter().handleCaptureVideo();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if (getConfig() == null) {
            return true;
        }

        if (getConfig().isEnableCamera()) {
            if (getConfig().isShowPhotos() && getConfig().isShowVideos()) {
                getMenuInflater().inflate(R.menu.mp_image_picker_menu_all_main, menu);
            } else if (getConfig().isShowVideos()) {
                getMenuInflater().inflate(R.menu.mp_image_picker_menu_video_main, menu);
            } else if (getConfig().isShowPhotos()) {
                getMenuInflater().inflate(R.menu.mp_image_picker_menu_photo_main, menu);
            }
        }
        return true;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        PickerActivityPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
    }

    public void startFiles(Dir dir) {
        new FilesActivity.Builder(this, dir)
                .start();
    }

    public void updateFiles() {
        if (getPresenter() != null) {
            getPresenter().updateFiles(new MediaFileLoader(this));
        }
    }

    public final void submitResult(List<MediaFile> files) {
        Intent intent = new Intent();

        final File result = new File(getCacheDir(), PickerConst.RESULT_FILE_NAME);

        try (FileOutputStream fos = new FileOutputStream(result, false)) {
            final Gson gson = new GsonBuilder().excludeFieldsWithoutExposeAnnotation().create();
            final String json = gson.toJson(files);
            fos.write(json.getBytes());
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        intent.setData(Uri.fromFile(result));

        setResult(RESULT_OK, intent);
        finish();
    }

    @SuppressWarnings("UnusedReturnValue")
    private boolean deleteConfig() {
        final File in = new File(getCacheDir(), DirsActivity.CONFIG_FILE_NAME);
        if (in.exists() && in.canWrite()) {
            return in.delete();
        }

        return false;
    }

    @NeedsPermission({Manifest.permission.CAMERA, Manifest.permission.WRITE_EXTERNAL_STORAGE})
    public void capturePhoto(CameraHandler cameraHandler, int requestCode) {
        Intent intent = cameraHandler.getCameraPhotoIntent(this, getConfig(), getCapturedSavePath());
        if (intent == null) {
            Toast.makeText(this, R.string.mp_error_create_image_file, Toast.LENGTH_LONG).show();
            return;
        }
        if (intent.resolveActivity(getPackageManager()) != null) {
            startActivityForResult(intent, requestCode);
        } else {
            Toast.makeText(this, R.string.mp_error_create_image_file, Toast.LENGTH_LONG).show();
        }
    }

    @NeedsPermission({Manifest.permission.CAMERA, Manifest.permission.WRITE_EXTERNAL_STORAGE})
    public void captureVideo(CameraHandler cameraHandler, int requestCode) {
        Intent intent = cameraHandler.getCameraVideoIntent(this, getConfig(), getCapturedSavePath());
        if (intent == null) {
            Toast.makeText(this, R.string.mp_error_create_video_file, Toast.LENGTH_LONG).show();
            return;
        }
        if (intent.resolveActivity(getPackageManager()) != null) {
            startActivityForResult(intent, requestCode);
        } else {
            Toast.makeText(this, R.string.mp_error_create_video_file, Toast.LENGTH_LONG).show();
        }
    }

    protected void setGridSpacingItemDecoration(RecyclerView list, int spanCount) {
        Timber.d("Set grid spacing");
        if (mGridSpacingItemDecoration != null) {
            list.removeItemDecoration(mGridSpacingItemDecoration);
        }
        mGridSpacingItemDecoration = new GridSpacingItemDecoration(spanCount, DisplayHelper.dpToPx(this, 1), false);
        list.addItemDecoration(mGridSpacingItemDecoration);
    }

    protected PickerSavePath getCapturedSavePath() {
        return null;
    }

    abstract protected PickerPresenter getPresenter();

    @Override
    public void rescanFiles(OnCompleteScan listener) {
        new Thread(() -> {
            Timber.d("Scan files...");
            scanFilesIn(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM).getAbsolutePath());
            scanFilesIn(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).getAbsolutePath());
            scanFilesIn(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).getAbsolutePath());
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                scanFilesIn(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS).getAbsolutePath());
            }
            if (listener != null) {
                new Handler(Looper.getMainLooper()).post(listener::onComplete);
            }
            scannedDirs = true;
        }).start();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == getConfig().getRequestCode() && resultCode == RESULT_OK) {
            setResult(resultCode, data);
            finish();
            return;
        }
        getPresenter().handleExtras(requestCode, resultCode, data);
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        if (getPresenter() != null) {
            getPresenter().onRestoreSavedState(savedInstanceState);
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        if (getPresenter() != null) {
            getPresenter().onSaveInstanceState(outState);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getPresenter() != null) {
            getPresenter().attachToLifecycle(this);
            getPresenter().onRestoreSavedState(savedInstanceState);
        }

        if (!scannedDirs) {
            rescanFiles();
        }

        setResult(RESULT_CANCELED);
    }

    private List<Pair<String, String>> iterateDirs(String pathName) {
        File folderFile = new File(pathName);
        File[] files = folderFile.listFiles();
        if (files != null) {
            List<Pair<String, String>> out = new ArrayList<>();
            for (File file : files) {
                // skip hidden files or directories
                if (file.getName().startsWith(".")) {
                    file = null;
                    continue;
                }

                // checking the File is file or directory
                if (file.isFile()) {
                    String mime = null;
                    String p = file.getAbsolutePath();
                    if (PickerUtils.isVideoFormat(p)) {
                        mime = PickerUtils.mimeFrom(p);
                    } else if (PickerUtils.isImageFormat(p)) {
                        mime = PickerUtils.mimeFrom(p);
                    }

                    if (mime != null) {
                        out.add(Pair.create(file.getAbsolutePath(), mime));
                    }
                    p = null;
                    file = null;
                    files = null;
                    folderFile = null;

                } else if (file.isDirectory()) {
                    out.addAll(iterateDirs(file.getAbsolutePath()));
                }
            }
            return out;
        }

        return Collections.emptyList();
    }

    private void scanFilesIn(String pathName) {
        List<Pair<String, String>> out = iterateDirs(pathName);
        List<List<Pair<String, String>>> chunks = chopped(out, 10);

        int pn = 0;
        for (List<Pair<String, String>> p : chunks) {
            String[] paths = new String[p.size()];
            String[] mimes = new String[p.size()];

            int i = 0;
            for (Pair<String, String> sp : p) {
                paths[i] = sp.first;
                mimes[i] = sp.second;
                i++;
            }

            MediaScannerConnection.scanFile(getApplicationContext(), paths, mimes, (path, uri) -> {
                Timber.d("Scanned: %s", path);
            });

            pn++;
        }
    }

    public interface OnCompleteScan {
        void onComplete();
    }

}

File: multipicker/src/main/java/com/edwardstock/multipicker/picker/ui/PickerActivity.kt
package com.edwardstock.multipicker.picker.ui


import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.media.MediaScannerConnection
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.addCallback
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContract
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.content.FileProvider
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentTransaction
import androidx.lifecycle.lifecycleScope
import androidx.transition.ChangeBounds
import androidx.transition.ChangeClipBounds
import androidx.transition.ChangeImageTransform
import androidx.transition.Fade
import androidx.transition.TransitionSet
import com.edwardstock.multipicker.PickerConfig
import com.edwardstock.multipicker.R
import com.edwardstock.multipicker.data.Dir
import com.edwardstock.multipicker.data.MediaFile
import com.edwardstock.multipicker.databinding.MpActivityPickerBinding
import com.edwardstock.multipicker.internal.MultiPickerFileProvider
import com.edwardstock.multipicker.internal.PickerSavePath
import com.edwardstock.multipicker.internal.PickerUtils
import com.edwardstock.multipicker.internal.views.ErrorView
import com.edwardstock.multipicker.picker.PickerConst
import com.edwardstock.multipicker.picker.getParcelableArrayListExtraCompat
import com.edwardstock.multipicker.picker.getParcelableCompat
import com.edwardstock.multipicker.picker.views.PickerView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import timber.log.Timber
import java.io.File
import java.lang.ref.WeakReference

/**
 * android-multipicker. 2018
 * @author Eduard Maximovich [edward.vstock@gmail.com]
 */

val capturePerms: Array<String> = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
    arrayOf(Manifest.permission.CAMERA)
} else {
    arrayOf(Manifest.permission.CAMERA, Manifest.permission.WRITE_EXTERNAL_STORAGE)
}

class PickerActivity : AppCompatActivity(), PickerView, ErrorView {
    lateinit var b: MpActivityPickerBinding
    private var config: PickerConfig? = null
    private var fragment: PickerFileSystemFragment? = null

    private var takePictureLauncher: ActivityResultLauncher<Uri>? = null
    private var takePictureFile: PickerSavePath? = null
    private var takeVideoLauncher: ActivityResultLauncher<Uri>? = null
    private var takeVideoFile: PickerSavePath? = null

    private enum class CaptureType {
        Photo, Video
    }

    private var captureType: CaptureType? = null

    private val launchPermissionRequest = registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) { permissions ->
        val granted = permissions.entries.all { it.value }
        if (granted) {
            when (captureType) {
                CaptureType.Photo -> capturePhotoInternal()
                CaptureType.Video -> captureVideoInternal()
                else -> Unit
            }
        } else {
            when (captureType) {
                CaptureType.Photo -> Toast.makeText(this, R.string.mp_error_create_image_file, Toast.LENGTH_LONG).show()
                CaptureType.Video -> Toast.makeText(this, R.string.mp_error_create_video_file, Toast.LENGTH_LONG).show()
                else -> Unit
            }
        }
    }

    override fun capturePhotoWithPermissions() {
        captureType = CaptureType.Photo
        launchPermissionRequest.launch(capturePerms)
    }

    override fun captureVideoWithPermissions() {
        captureType = CaptureType.Video
        launchPermissionRequest.launch(capturePerms)
    }

    private fun capturePhotoInternal() {
        try {
            takePictureFile = PickerSavePath.newTimestampTmpFile(this, "jpg")
            takePictureLauncher?.launch(takePictureFile!!.toUriProvider(this))
        } catch (t: Throwable) {
            Timber.e(t, "Unable to take picture")
            Toast.makeText(this, R.string.mp_error_create_image_file, Toast.LENGTH_LONG).show()
        }
    }

    private fun captureVideoInternal() {
        try {
            takeVideoFile = PickerSavePath.newTimestampTmpFile(this, "mp4")
            takeVideoLauncher?.launch(takeVideoFile!!.toUriProvider(this))
        } catch (t: Throwable) {
            Timber.e(t, "Unable to take video")
            Toast.makeText(this, R.string.mp_error_create_video_file, Toast.LENGTH_LONG).show()
        }
    }

    private fun onCapturedPhoto(saved: Boolean) {
        if (!saved) {
            Toast.makeText(this, R.string.mp_error_create_image_file, Toast.LENGTH_LONG).show()
        } else {
            val file = takePictureFile!!.toFile(this)!!
            val uri = FileProvider.getUriForFile(this, MultiPickerFileProvider.getAuthority(this), file)
            val mf = MediaFile(
                    0,
                    file.name,
                    file.absolutePath,
                    uri
            )
            try {
                PickerUtils.writeCapturedMedia(this, mf)
                lifecycleScope.launch {
                    takePictureFile = null
                    startUpdateFiles()
                }
            } catch (e: Throwable) {
                Timber.e(e)
                Toast.makeText(this, R.string.mp_error_create_image_file, Toast.LENGTH_LONG).show()
            }
        }
    }

    private fun onCapturedVideo(captured: Boolean) {
        if (captured) {
            val file = takeVideoFile!!.toFile(this)!!
            val uri = FileProvider.getUriForFile(this, MultiPickerFileProvider.getAuthority(this), file)

            val mf = MediaFile(
                    0,
                    file.name,
                    file.absolutePath,
                    uri
            )
            if (!mf.uri.exists() || mf.uri.length() == 0L) {
                Toast.makeText(this, R.string.mp_error_create_video_file, Toast.LENGTH_LONG).show()
                Timber.w("Unable to handle recorded video: file not exists or is empty")
                return
            }

            try {
                PickerUtils.writeCapturedMedia(this, mf)
                lifecycleScope.launch {
                    takeVideoFile = null
                    startUpdateFiles()
                }
            } catch (e: Throwable) {
                Timber.e(e)
                Toast.makeText(this, R.string.mp_error_create_video_file, Toast.LENGTH_LONG).show()
            }
        } else {
            Toast.makeText(this, R.string.mp_error_create_video_file, Toast.LENGTH_LONG).show()
        }

    }

    override fun scanMedia(file: MediaFile, listener: MediaScannerConnection.OnScanCompletedListener) {
        val paths = arrayOf(file.uri)
        val mimeTypes = arrayOf(if (file.isVideo) "video/mp4" else "image/jpeg")
        MediaScannerConnection.scanFile(applicationContext, paths, mimeTypes, listener)
    }

    override fun onError(error: CharSequence?) {
        Toast.makeText(this, error, Toast.LENGTH_LONG).show()
    }

    override fun onError(t: Throwable?) {
        Toast.makeText(this, t?.message, Toast.LENGTH_LONG).show()
    }

    override fun startUpdateFiles() {
        if (fragment != null && fragment?.isAdded == true) {
            fragment?.updateFiles()
        }
    }

    fun setupToolbar(toolbar: Toolbar) {
        setSupportActionBar(toolbar)
        supportActionBar?.setHomeButtonEnabled(true)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressedDispatcher.onBackPressed()
            return true
        }
        if (item.itemId == R.id.menu_camera) {
            capturePhotoWithPermissions()
            return true
        } else if (item.itemId == R.id.menu_video) {
            captureVideoWithPermissions()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        if (config == null) {
            return true
        }
        if (config!!.isEnableCamera) {
            if (config!!.isShowPhotos && config!!.isShowVideos) {
                menuInflater.inflate(R.menu.mp_image_picker_menu_all_main, menu)
            } else if (config!!.isShowVideos) {
                menuInflater.inflate(R.menu.mp_image_picker_menu_video_main, menu)
            } else if (config!!.isShowPhotos) {
                menuInflater.inflate(R.menu.mp_image_picker_menu_photo_main, menu)
            }
        }
        return true
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        onBackPressedDispatcher.addCallback(this, true) {
            handleBackPress()
        }
        config = intent?.getParcelableCompat(PickerConst.EXTRA_CONFIG)
        b = MpActivityPickerBinding.inflate(layoutInflater)
        setContentView(b.root)
        setResult(RESULT_CANCELED)
        setupToolbar(b.toolbar)
        b.toolbar.setNavigationOnClickListener {
            handleBackPress()
        }

        if (config?.title != null) {
            b.toolbar.title = config?.title
        }

        if (config?.isShowPhotos == true && config?.isShowVideos == true) {
            takePictureLauncher = registerForActivityResult(ActivityResultContracts.TakePicture(), this::onCapturedPhoto)
            takeVideoLauncher = registerForActivityResult(ActivityResultContracts.CaptureVideo(), this::onCapturedVideo)
        } else if (config?.isShowVideos == true) {
            takeVideoLauncher = registerForActivityResult(ActivityResultContracts.CaptureVideo(), this::onCapturedVideo)
        } else if (config?.isShowPhotos == true) {
            takePictureLauncher = registerForActivityResult(ActivityResultContracts.TakePicture(), this::onCapturedPhoto)
        }

        if (savedInstanceState == null) {
            // If there is no saved instance state, create a new Fragment
            fragment = DirsFragment.newInstance(config)
            Timber.d("Show DirsFragment")
            supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.mp_container, fragment!!, "files_fragment")
                    .addToBackStack(null)
                    .commit()
        } else {
            // If there is a saved instance state, get the Fragment by its tag
            fragment = supportFragmentManager.findFragmentByTag("files_fragment") as PickerFileSystemFragment
            Timber.d("Show ${fragment!!::class.simpleName} fragment from saved instance state")
        }
    }

    private fun handleBackPress() {
        fragment?.onBackPressed()
        var addedFile: MediaFile? = null
        if (fragment is ImageViewerFragment) {
            addedFile = (fragment as ImageViewerFragment).addedFile
        }
        if (supportFragmentManager.backStackEntryCount <= 1) {
            setResult(RESULT_CANCELED)
            finish()
        } else {
            resolveLastFragment()
            if (fragment is FilesFragment) {
                (fragment as FilesFragment).addSelection(addedFile)
            }
            supportFragmentManager.popBackStack()
        }
    }

    fun startFiles(dir: Dir?) {
        val fragment = FilesFragment.newInstance(config, dir)
        Timber.d("Show FilesFragment")
        supportFragmentManager
                .beginTransaction()
                .replace(R.id.mp_container, fragment, "files_fragment")
                .addToBackStack(null)
                .commit()
        this.fragment = fragment
    }

    fun startPreview(dir: Dir?, file: MediaFile, sharedView: View?) {
        if (file.isVideo) {
            val intent = Intent(Intent.ACTION_VIEW, file.uri)
            intent.setDataAndType(file.uri, "video/*")
            startActivity(intent)
        } else {
            val fragment = ImageViewerFragment.newInstance(dir, file)
            val tx: FragmentTransaction = supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.mp_container, fragment, fragment.javaClass.simpleName)

            if (sharedView != null) {
                tx.addSharedElement(sharedView, sharedView.transitionName)
                val sharedSet = TransitionSet()
                sharedSet.addTransition(ChangeImageTransform())
                sharedSet.addTransition(ChangeBounds())
                sharedSet.addTransition(ChangeClipBounds())
                //                sharedSet.addTransition(new ChangeTransform());
                val commonSet = TransitionSet()
                commonSet.addTransition(Fade())
                fragment.sharedElementEnterTransition = sharedSet
                fragment.sharedElementReturnTransition = sharedSet
                fragment.enterTransition = commonSet
                fragment.returnTransition = commonSet
            }
            tx.addToBackStack(null).commit()
            this.fragment = fragment
        }
    }

    fun submitResult(files: List<MediaFile>) {
        val intent = Intent()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && config?.copyUnreadableFilesToCache == true) {
            lifecycleScope.launch(Dispatchers.IO) {
                files
                        .filter { !it.uri.canRead() }
                        .forEach { source ->
                            val mediaPath = File(externalCacheDir, MultiPickerFileProvider.EXT_CACHE_DIR).also { it.mkdirs() }
                            val mediaFile = File(mediaPath, source.name)

                            // if target file does not exists OR it does but length isn't the same
                            if (mediaFile.exists().not() || mediaFile.length() != source.size) {
                                mediaFile.delete()
                                Timber.d("Copy ${source.uri} to local cache")

                                mediaFile.createNewFile()
                                contentResolver.openInputStream(source.uri)?.use { inputStream ->
                                    mediaFile.outputStream().use { outputStream ->
                                        inputStream.copyTo(outputStream)
                                    }
                                }

                                val contentUri: Uri = FileProvider.getUriForFile(this@PickerActivity, MultiPickerFileProvider.getAuthority(this@PickerActivity), mediaFile)
                                MultiPickerFileProvider.grantFilePermissions(this@PickerActivity, contentUri, true)
                                source.uri = mediaFile.absolutePath
                            }
                        }

                withContext(Dispatchers.Main) {
                    intent.putParcelableArrayListExtra(PickerConst.EXTRA_RESULT_FILES, ArrayList(files))
                    setResult(RESULT_OK, intent)
                    finish()
                }
            }
        } else {
            intent.putParcelableArrayListExtra(PickerConst.EXTRA_RESULT_FILES, ArrayList(files))
            setResult(RESULT_OK, intent)
            finish()
        }

    }


    private fun checkFiles(pathName: String) {
        val folderFile = File(pathName)
        val files = folderFile.listFiles()
        if (files != null) {
            for (file in files) {
                // checking the File is file or directory
                if (file.isFile) {
                    val extension = file.absolutePath.substring(file.absolutePath.lastIndexOf(".") + 1)
                    var mime: String? = null
                    if (extension.equals("mp4", ignoreCase = true)) {
                        mime = "video/mp4"
                    } else if (extension.equals("jpg", ignoreCase = true)) {
                        mime = "image/jpeg"
                    }
                    if (mime != null) {
                        Timber.d("Scan file: %s", file.absolutePath)
                        MediaScannerConnection.scanFile(applicationContext, arrayOf(file.parent), arrayOf(mime), null)
                    }
                } else if (file.isDirectory) {
                    checkFiles(file.absolutePath)
                }
            }
        }
    }

    private fun resolveLastFragment() {
        val index: Int = supportFragmentManager.backStackEntryCount - 1
        if (index >= 0) {
            val backEntry: FragmentManager.BackStackEntry = supportFragmentManager.getBackStackEntryAt(index)
            val tag = backEntry.name
            if (tag != null) {
                fragment = supportFragmentManager.findFragmentByTag(tag) as PickerFileSystemFragment
            }
        }
    }

    class PickMediaFiles : ActivityResultContract<PickerConfig, List<MediaFile>>() {
        override fun createIntent(context: Context, input: PickerConfig): Intent {
            val intent = Intent(context, PickerActivity::class.java)
            intent.putExtra(PickerConst.EXTRA_CONFIG, input)
            return intent
        }

        override fun parseResult(resultCode: Int, intent: Intent?): List<MediaFile> {
            val emptyRes = emptyList<MediaFile>()
            if (resultCode == Activity.RESULT_OK) {
                if (intent == null) {
                    return emptyRes
                }

                return intent.getParcelableArrayListExtraCompat(PickerConst.EXTRA_RESULT_FILES) ?: emptyRes
            }
            return emptyRes
        }

    }


    /**
     * Uses new ActivityResult API, so after prepare, you should take result of {@link Builder#prepare(onResult: (List<MediaFile>) -> Unit)} method and then
     * call {@link ActivityResultLauncher#launch(PickerConfig)}
     */
    class Builder {
        private var activity: WeakReference<ComponentActivity>? = null
        private var fragment: WeakReference<Fragment>? = null

        constructor(from: ComponentActivity) {
            activity = WeakReference(from)
        }

        constructor(from: Fragment) {
            fragment = WeakReference(from)
        }

        fun prepare(onResult: (List<MediaFile>) -> Unit): ActivityResultLauncher<PickerConfig> {
            return when {
                activity != null && activity?.get() != null -> {
                    activity!!.get()!!.registerForActivityResult(PickMediaFiles(), onResult)
                }
                fragment != null && fragment?.get() != null -> {
                    fragment!!.get()!!.registerForActivityResult(PickMediaFiles(), onResult)
                }
                else -> {
                    throw IllegalStateException("Unable to start MultiPicker as source Activity or Fragment is null")
                }
            }
        }
    }
}